RedHatInsights/rhsm-subscriptions
# ENT-5154 Send TallySummary messages after snapshots are persisted

https://issues.redhat.com/browse/ENT-5154

Remittance logic requires that applicable snapshots be persisted before the
resulting TallySummary message is consumed. When processing a lot of events that result in
a lot of snapshots (for example during a re-tally), message consumption can sometimes be a little
of snapshot persistance. This can result in an incorrect calculated remittance.

Moved the message sending outside of the DB transaction. When remittance
is updated, we require that all relevant snapshots be persisted so that
our totals are accurate since TallySummary messages can be received out
of order.

## Testing

1. Clean out the applicable data
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "delete from events; delete from tally_snapshots; delete from instance_monthly_totals; delete from instance_measurements; delete from billable_usage_remittance;"
```

2. Load events into the DB. Make sure that it includes Events for the metrics that we are switching over (Reach out for sql script to load events if you need some, or you can't reproduce with events that you already have.)
```
psql -U rhsm-subscriptions rhsm-subscriptions < ~/events_remittance_update.sql
```

3. Check out the develop branch and deploy the application
```
SPRING_PROFILES_ACTIVE="worker,api,kafka-queue" DEV_MODE=true ./gradlew :bootRun
```

4. Make sure that you have opted in all accounts relevant to your events.
```
curl 'http://localhost:9000/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=optInJmxBean,type=OptInJmxBean","operation":"createOrUpdateOptInConfig(java.lang.String,java.lang.String,boolean,boolean,boolean)","arguments":["6043640", "333333", true, true, true]}'
```

5. Do an hourly tally for all accounts.
```
curl 'http://localhost:9000/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAllAccountsByHourly(java.lang.String,java.lang.String)","arguments":["2022-06-01T00:00Z", "2022-07-01T00:00Z"]}'
```

6. Check the remittance table after the hourly tally.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select * from billable_usage_remittance order by account_number, product_id, metric_id, billing_provider;"
```
```
 account_number | org_id | product_id |        metric_id        | accumulation_period |   sla   |   usage    | billing_provider | billing_account_id | remitted_value |        remittance_date        | version 
----------------+--------+------------+-------------------------+---------------------+---------+------------+------------------+--------------------+----------------+-------------------------------+---------
 6043640        |        | rhosak     | Instance-hours          | 2022-06             | Premium | Production | aws              | 027933477046       |            692 | 2022-07-04 14:01:31.728989+00 |       8
 6043640        |        | rhosak     | Storage-gibibyte-months | 2022-06             | Premium | Production | aws              | 027933477046       |             37 | 2022-07-04 14:01:29.817753+00 |       4
 6043640        |        | rhosak     | Transfer-gibibytes      | 2022-06             | Premium | Production | aws              | 027933477046       |              1 | 2022-07-04 14:01:29.64637+00  |       0
(3 rows)

```
7. Run the hourly tally again, and the remitted value will change.
```
curl 'http://localhost:9000/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAllAccountsByHourly(java.lang.String,java.lang.String)","arguments":["2022-06-01T00:00Z", "2022-07-01T00:00Z"]}'
```
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select * from billable_usage_remittance order by account_number, product_id, metric_id, billing_provider;"
```
```
 account_number | org_id | product_id |        metric_id        | accumulation_period |   sla   |   usage    | billing_provider | billing_account_id | remitted_value |        remittance_date        | version 
----------------+--------+------------+-------------------------+---------------------+---------+------------+------------------+--------------------+----------------+-------------------------------+---------
 6043640        |        | rhosak     | Instance-hours          | 2022-06             | Premium | Production | aws              | 027933477046       |            695 | 2022-07-04 14:03:20.417856+00 |       9
 6043640        |        | rhosak     | Storage-gibibyte-months | 2022-06             | Premium | Production | aws              | 027933477046       |             38 | 2022-07-04 14:03:21.481604+00 |       5
 6043640        |        | rhosak     | Transfer-gibibytes      | 2022-06             | Premium | Production | aws              | 027933477046       |              1 | 2022-07-04 14:01:29.64637+00  |       0
(3 rows)
```

8. Deploy this branch, and repeat all steps. Note that the remitted value after both tally runs should be the same (695 in my example).
